Student: submit responses: show a 'closing soon' warning #1872

HirdayGupta opened the pull request:
Fixes #1872 

**Outline of Solution:**

In `feedbackSubmissionsEdit.es6`, run a check in `document.ready()` if the FS is about to close, and accordingly either display "Closing Soon" warning modal or not. 

It follows a similar approach to @AnaghHegde's in #7179, but there is **a crucial difference:** 

**#7179's approach:**
In #7179, each element of the date was checked (ie: month vs month, year vs year, hours vs hours, etc). It checked if all elements of the current date and end date are the same (year, month, day, **hour**) and then checks if the difference between minutes is less than 15. This approach would lead to the feature not working as expected in the following cases (assuming year, month, date to be same) and some severe complications in fixing it for the following:

| Current Time  | End Time for Session | Result
| ------------- | ------------- | ------------- |
| 13:50  | 14:00  | No modal displayed even though only 10 mins remaining as the hour is not the same |
| 14:10  | 14:00  | Modal of "closing soon" displayed even though session has ended. |

**My approach:**
In my approach, I simply subtract the two date objects. This gives me the difference between the current datetime and end datetime in milliseconds. Then, I simply check if this difference is less than `900000 ms (ms in 15 mins)` AND greater than 0 (so that 'closing soon' is not displayed after session is closed). In my opinion, this approach is cleaner, more readable and also prevents the unnecessary complications that arise from the previously discussed approach.

**Input required from Senior Devs:**
- Currently, the modal displays ONLY IF the page is loaded when there are less than 15 minutes remaining. I was wondering: would it be overkill to add a timer event that runs every pre-defined time interval (say 5 minutes) and displays the warning when there's fifteen minutes left? This way even if someone loaded the page before the session was about to close would be able to get the warning. To optimise we could add the timer event only if the page was loaded with say, an hour to go.

- Also, the modal simply reads "Warning: you have **less than 15 minutes remaining**..." and displays when there's less than 15 mins left. Should I make it so that instead of simply saying you have "less than 15 mins" it displays the exact amount of time left like "you have x minutes remaining"? Or would that, again, be overkill?

HirdayGupta commented:
Dropping a note (again) that the only failing test on Travis is passing locally. Requesting to restart the build. :)

whipermr5 commented on src/main/webapp/dev/js/main/feedbackSubmissionsEdit.es6:
```diff
@@ -919,6 +923,15 @@ function hasWarningMessage() {
 }
 
+function isSessionClosingSoon() {
+    const endDate = new Date($(END_TIME).html().trim());
```
Does this handle time zone?

HirdayGupta replied:
Will add this information to the page in #7736 and then handle in time zone here.

whipermr5 commented on src/main/webapp/dev/js/main/feedbackSubmissionsEdit.es6:
```diff
@@ -919,6 +923,15 @@ function hasWarningMessage() {
+function isSessionClosingSoon() {
+    const endDate = new Date($(END_TIME).html().trim());
+    const currentDate = new Date();
+    if (endDate - currentDate <= MS_IN_FIFTEEN_MINUTES && endDate - currentDate > 0) {
```
`endDate - currentDate` can be extracted into a variable `remainingTime` since it's used twice.

whipermr5 commented:
Yes; this approach is correct 👍

> Currently, the modal displays ONLY IF the page is loaded when there are less than 15 minutes remaining. I was wondering: would it be overkill to add a timer event that runs every pre-defined time interval (say 5 minutes) and displays the warning when there's fifteen minutes left? This way even if someone loaded the page before the session was about to close would be able to get the warning. To optimise we could add the timer event only if the page was loaded with say, an hour to go.
>
> Also, the modal simply reads "Warning: you have less than 15 minutes remaining..." and displays when there's less than 15 mins left. Should I make it so that instead of simply saying you have "less than 15 mins" it displays the exact amount of time left like "you have x minutes remaining"? Or would that, again, be overkill?

Perhaps there can be a visible countdown timer on the page if the time remaining is an hour or less. When it crosses the 15 minutes remaining mark, the modal can be triggered once.

damithc commented:
> Perhaps there can be a visible countdown timer on the page if the time remaining is an hour or less. When it crosses the 15 minutes remaining mark, the modal can be triggered once.

This is a good idea, but I'd rather not do it. The reason is because some sessions have a hidden grace period which is not known to students, so we don't want to be too precise about when the deadline expires. :-p

HirdayGupta commented:
Hi @whipermr5. Added in your suggestion to extract `endDate - currentDate` to another `const`. Also, based on above discussion, I'm assuming the current approach of displaying a "Warning: You have less than 15 minutes remaining..." modal is the ideal one. Please correct me if I'm wrong. 

Except for the timezone feature, which I am working on in #7736, this PR is ready for review. I will add the timezone functionality in as soon as the PR for #7736 is merged.

whipermr5 commented on src/main/webapp/WEB-INF/tags/shared/feedbackSubmissionEdit/feedbackSessionDetailsPanel.tag:
```diff
@@ -30,7 +30,7 @@
                     <label class="col-sm-2 control-label">Closing time:</label>
                     <div class="col-sm-10">
-                        <p class="form-control-static">${feedbackSession.endTimeString}</p>
+                        <p class="form-control-static" id="end-time">${feedbackSession.endTimeString}</p>
```
I'd prefer if the UTC time is embedded as a data attribute here in [JavaScript-friendly format](https://stackoverflow.com/questions/10286204/the-right-json-date-format?answertab=votes#tab-top), e.g.
```
<p class="form-control-static" id="end-time" data-end-time="2012-04-23T18:25:43.511Z">Tue, 24 Apr 2012, 2:25 AM UTC+8</p>
```
This would make the data independent of the way it's displayed. Also, the time zone info wouldn't have to be processed from JavaScript. Instead, the date object can be created simply with:
```
const endDate = new Date($(END_TIME).data('endTime'));
```

HirdayGupta replied:
@whipermr5 How would I get the date in that format though? Right now the date is in a very different format.

whipermr5 replied:
The format is called ISO 8601. Write a method in `TimeHelper` if none of the existing methods are relevant.

whipermr5 replied:
This change will have to be done in #7740.

HirdayGupta replied:
In the second approach, a benefit is that when I parse the date using JS's Date class, it automatically converts it to the local timezone and then the comparison between current date and end date is exactly as intended.

whipermr5 replied:
UTC but with the `Z` suffix. That way the conversion should be automatic too.

HirdayGupta replied:
Ahh okay, I see. Any suggestions on converting to UTC in time helper? The `endTime` `java.util.Date` object from `FeedbackSessionAttributes` contains no timezone information at all so this will be slighty tricky.

whipermr5 replied:
After looking through `TimeHelper`, the `convertToUserTimeZone` method seems useful here :)

HirdayGupta replied:
@whipermr5 Thanks for that, that was very helpful! Apologies for not picking that up myself, I am a little new to Java's Date and Calendar classes and as such was unsure what exactly everything was doing :p. Have read up on the docs now, and should, hopefully, not need as much hand-holding. Thanks again! 😄

HirdayGupta commented:
@whipermr5 Why the `s.OnHold` label?

whipermr5 commented:
Waiting for #7740 to be merged :)

HirdayGupta commented:
Ahh okay 😅 

But in your suggestion above, you had recommended that I make this PR's parsing of date independent of how the date is displayed. So here, I would focus on getting the correctly formatted time into `data-end-time` and then in #7740, I format the way it is displayed?

whipermr5 commented:
Yupp sounds good!

HirdayGupta commented:
@whipermr5 added the `data-end-time` attribute in UTC timezone is ISO 8601 format as requested. Could you perhaps do a quick review of 6dc6e37 and let me know if any changes required?

whipermr5 commented on src/main/java/teammates/common/util/TimeHelper.java:
```diff
@@ -252,6 +252,16 @@ public static String formatDateTimeForInstructorHomePage(Date date) {
         return sdf.format(date);
     }
 
+    public static String formatDateForFeedbackSubmissionEditPage(Date date, double timeZone) {
```
This method name should be generalised to ISO8601 instead.

whipermr5 commented on src/main/java/teammates/common/util/TimeHelper.java:
```diff
@@ -252,6 +252,16 @@ public static String formatDateTimeForInstructorHomePage(Date date) {
+        }
+        // Passing negative timeZone to convert to UTC.
+        Calendar calInUtc = convertToUserTimeZone(dateToCalendar(date), -timeZone);
+        SimpleDateFormat sdf = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.S'Z'");
```
This can be further extracted into a helper method. Extract the string into Constants.

whipermr5 commented on src/main/java/teammates/common/util/TimeHelper.java:
```diff
@@ -252,13 +252,13 @@ public static String formatDateTimeForInstructorHomePage(Date date) {
     }
 
-    public static String formatDateForFeedbackSubmissionEditPage(Date date, double timeZone) {
+    public static String formateDateToIso8601(Date date, double timeZone) {
```
There's a typo in the method name. Also, move the conversion to UTC logic into a separate method, as from the method name, this method should do nothing other than conversion.

HirdayGupta commented:
@whipermr5 
 - Generalised method name to `formateDateToIso8601`
- Extracted the ISO8601 Format String to `Const.TIME_FORMAT_ISO_8601`
- Added a javadoc comment for the `formateDateToIso8601` method. (Not requested, but I felt it necessary).

For your comment on `Line 261`, where you said "This can be further extracted into a helper method", I'm not sure exactly what you meant. Could you clarify further? Should I extract the line that formats the date to ISO8601 to another method? What would be the benefit of this?

Other than the above clarification, all changes made and **ready for review**.

whipermr5 commented:
@HirdayGupta See https://github.com/TEAMMATES/teammates/pull/7715#discussion_r127649000. Basically, the method shouldn't do more than one thing.

HirdayGupta commented:
@whipermr5 I see your point. However, seeing that in the ISO8601 format, a `Z` is appended to the end of the string indicating that the date is in UTC, I think that converting the date to UTC is a part of the process of converting to ISO8601 format. And, if the date is already in UTC, then whoever wishes to use the helper method can simply pass 0.0 as the timeZone.

whipermr5 commented:
> I think that converting the date to UTC is a part of the process of converting to ISO8601 format

That's fair, but I believe the ISO8601 standard doesn't actually specify which time zone to use. Maybe rename the method to `convertAndFormatDateToIso8601Utc`?

HirdayGupta commented:
> Maybe rename the method to convertAndFormatDateToIso8601Utc?

Yep that sounds good. Done @whipermr5. Will update the expected HTML test files using god mode soon. Anything else I should change in my implementation?

whipermr5 commented on src/main/java/teammates/common/util/Const.java:
```diff
@@ -110,6 +110,8 @@
         TIME_REPRESENTS_DEFAULT_TIMESTAMP = TimeHelper.convertToDate("2011-01-01 00:00 AM UTC");
     }
 
+    public static final String TIME_FORMAT_ISO_8601 = "yyyy-MM-dd'T'HH:mm:ss.S'Z'";
```
Rename to `TIME_FORMAT_ISO_8601_UTC` here too.

whipermr5 commented on src/main/webapp/WEB-INF/tags/shared/feedbackSubmissionEdit/feedbackSessionDetailsPanel.tag:
```diff
@@ -34,7 +34,7 @@
           <label class="col-sm-2 control-label">Closing time:</label>
           <div class="col-sm-10">
-            <p class="form-control-static">${feedbackSession.endTimeString}</p>
+            <p class="form-control-static" id="end-time" data-end-time="<%= feedbackSession.getEndTimeInIso8601Format() %>"  >${feedbackSession.endTimeString}</p>
```
`data-end-time="${feedbackSession.endTimeInIso8601Format}"`

HirdayGupta replied:
Whoa I had no clue this would work! How does this work? Does JSP automatically detect `getter` methods and create local attributes?

whipermr5 replied:
Yup it does! I learned recently haha, while working on #7385.

HirdayGupta replied:
That is pretty cool! Thanks again @whipermr5! (Also, #7385 looks super intense xD)

whipermr5 commented:
> Will update the expected HTML test files using god mode soon. Anything else I should change in my implementation?

Nothing else; you can go ahead :)

whipermr5 commented on src/main/webapp/WEB-INF/tags/shared/feedbackSubmissionEdit/feedbackSessionDetailsPanel.tag:
```diff
@@ -34,7 +34,7 @@
           <label class="col-sm-2 control-label">Closing time:</label>
           <div class="col-sm-10">
-            <p class="form-control-static">${feedbackSession.endTimeString}</p>
+            <p class="form-control-static" id="end-time" data-end-time="${feedbackSession.endTimeInIso8601Format}"  >${feedbackSession.endTimeString}</p>
```
Oops just noticed; delete the two spaces before the `>`.

whipermr5 replied:
Okay it's good to go now :)

HirdayGupta commented:
Dropping a note that I've looked into ba63c25 and it seems okay, as there is no end time for the session being tested in that case.

HirdayGupta commented:
@whipermr5 Not entirely sure but I think the build is failing because the end time for some reason changed between the time I ran the tests on my local machine and they time they ran on travis. I wonder why?

HirdayGupta commented:
Just reran the tests locally and the build passed (godmode was disabled). I am not too sure about whats going wrong here.

whipermr5 commented:
Somehow `feedbackSession.endTimeInIso8601Format` is not being output in UTC but in your local time zone.

HirdayGupta commented:
>Somehow feedbackSession.endTimeInIso8601Format is not being output in UTC but in your local time zone.

@whipermr5 on manually loading the page and viewing the source this doesn't seem to be the case. 

Just tried running some of the failing tests in eclipse and they generated the correct output as required by the Travis build. Seems like there is some problem with my configuration of `./gradlew appengineRun CiTests`.

HirdayGupta commented:
@whipermr5 Take a look at 8bc51f1 and the one failing test in [this build](https://travis-ci.org/TEAMMATES/teammates/builds/254440375?utm_source=github_status&utm_medium=notification). I think the expected value for the file that was changed in 8bc51f1 changes for each test run. What can I do about this?

whipermr5 commented:
@HirdayGupta You need to update `HtmlHelper` to detect the current time in ISO8601 format and replace it with a placeholder e.g. `${datetime.now.iso8601}`.

HirdayGupta commented:
@whipermr5 Couple of questions:

- Would it be enough for me to add a replaceAll() in the `replaceUnpredictableValuesWithPlaceholders` for the new format?

- Also, check out this SO answer: https://stackoverflow.com/questions/28020805/regex-validate-correct-iso8601-date-string-with-time. The regex for ISO 8601 seems MONSTROUS! Is there some other way to do this?

whipermr5 commented:
@HirdayGupta Check out how the `datetime.now` detection is implemented. Instead of using a regex, compute the current time in UTC in the correct format and find it via exact match :)

whipermr5 commented:
> Dropping a note that I've looked into ba63c25 and it seems okay, as there is no end time for the session being tested in that case.

How will this affect the JavaScript?

whipermr5 commented:
> EDIT: Everything else seems to be fixed by running the tests on Eclipse instead of on the Terminal. (I think that needs some investigating as well?)

Not sure if this is related to the changes I made in #7521. Do your Eclipse launch configurations have the UTC flag?

HirdayGupta commented:
> Not sure if this is related to the changes I made in #7521. Do your Eclipse launch configurations have the UTC flag?

@whipermr5 yes, my main `run as web app` run config has the UTC flag VM argument.

> How will this affect the JavaScript?

Will look into this and let you know soon!

> Check out how the datetime.now detection is implemented.

I believe it is done by getting the current date and then using a Regex to allow any value for `hh:mm`. (See [this line](https://github.com/TEAMMATES/teammates/blob/master/src/test/java/teammates/test/driver/HtmlHelper.java#L447)) I think I should follow a similar approach? So instead of using just `replace` like you suggested I will use `replaceAll` but only regex a part of the matching (ie hours and minutes), leaving the rest to absolute matching instead. Does this seem okay?

HirdayGupta commented:
> How will this affect the JavaScript?

So essentially, if the parameter on the date constructor is an empty string (which will be the case if the end date is empty), the `endDate` object gets initialised to `Invalid Date` and the `remainingTime` becomes `NaN`. As a result of this, no modal is displayed, since [all comparisons with NaN return false in JavaScript](https://stackoverflow.com/questions/26982881/why-nan-is-greater-than-any-number-in-javascript). So I think there is no need of any special handling of this situation.

HirdayGupta commented:
Also, @whipermr5 build has passed and the PR is ready fpr review! :-)

whipermr5 commented:
> yes, my main run as web app run config has the UTC flag VM argument.

Does it stop working if you remove the flag? Are you on the latest AppEngine SDK (1.9.54; #7627)?

> I believe it is done by getting the current date and then using a Regex to allow any value for hh:mm. (See this line)

Oops; I didn't realise that! I guess it's probably to avoid the situation where the time changes between the preparation of the value to match and the actual matching. Then again, the same thing can happen if the test is run close to midnight and the date changes halfway through. Could you try going with the complete exact match approach first just to see if it works?

> So essentially, if the parameter on the date constructor is an empty string (which will be the case if the end date is empty), the endDate object gets initialised to Invalid Date and the remainingTime becomes NaN. As a result of this, no modal is displayed, since all comparisons with NaN return false in JavaScript. So I think there is no need of any special handling of this situation.

I see; it sounds hackish to rely on that mechanism though. Writing this would make it more readable, as someone reading the code would immediately understand that the end time data can sometimes be empty:
```javascript
function isSessionClosingSoon() {
    const endTimeData = $(END_TIME).data('end-time');
    if (!endTimeData) {
        return false; // session has no end time
    }
    const endDate = new Date(endTimeData);
    ...
}
```

whipermr5 commented on src/main/webapp/dev/js/main/feedbackSubmissionsEdit.es6:
```diff
@@ -919,6 +923,16 @@ function hasWarningMessage() {
+    if (remainingTime <= MS_IN_FIFTEEN_MINUTES && remainingTime > 0) {
+        return true;
+    }
+    return false;
```
These four lines can be summarised to
```javascript
return remainingTime <= MS_IN_FIFTEEN_MINUTES && remainingTime > 0;
```

whipermr5 commented on src/main/java/teammates/common/util/TimeHelper.java:
```diff
@@ -251,6 +251,19 @@ public static String formatDateTimeForInstructorHomePage(Date date) {
+            return "";
+        }
+        // Passing negative timeZone to convert to UTC.
+        Calendar calInUtc = convertToUserTimeZone(dateToCalendar(date), -timeZone);
```
I believe you can now make use of your newly created method `convertLocalDateToUtc` :)

HirdayGupta replied:
@whipermr5 While I like the idea, the `convertLocalDateToUtc` method _is_ deprecated whereas this method is super useful and can live on even after `convertLocalDateToUtc` is removed. Given this, would it still be a good idea to use it?

HirdayGupta replied:
Ohh okay wait nevermind just read your comment below. This now makes sense because I will  be using the `convertLocalDateToUtc` method in `FeedbackSessionAttributes` and then it can be removed once we migrate all dates to UTC. 👍

whipermr5 commented on src/main/java/teammates/common/datatransfer/attributes/FeedbackSessionAttributes.java:
```diff
@@ -684,4 +684,8 @@ public void setRespondingInstructorList(Set<String> respondingInstructorList) {
     }
+
+    public String getEndTimeInIso8601Format() {
+        return TimeHelper.convertAndFormatDateToIso8601Utc(endTime, timeZone);
```
Actually on second thoughts, I still feel that the conversion and the formatting should be separated (follow how you did `getStartTimeString` and `getEndTimeString`). The method that formats to ISO8601 shouldn't do the conversion.

whipermr5 commented on src/main/java/teammates/common/util/TimeHelper.java:
```diff
@@ -266,14 +266,13 @@ public static String formatDateTimeForInstructorHomePage(Date date) {
     /**
      * Converts {@code date} to UTC according to the {@code timeZone} offset and formats it according to ISO8601.
```
Remember to update the javadoc too.

HirdayGupta commented:
@whipermr5 There is definitely a difference between running the tests using `./gradlew appengineRun ciTests` and running them from Eclipse. Just experienced once more where the test files were updated with the wrong time values when I used the Terminal method.

> Does it stop working if you remove the flag? Are you on the latest AppEngine SDK (1.9.54; #7627)?

**Yes, it stops working if I remove the flag**. (Wondering if this indicates something missed in #7521 or  something wrong with my local set up). Note: The values differ from the correct values by the offset of my local timezone (India - UTC + 5:30) from UTC. So `15:59` becomes `21:29` and so on. 

And yes, I am on the latest AppEngine SDK.

> Could you try going with the complete exact match approach first just to see if it works?

Yep. Will try this and let you know soon. However, your point about the close to midnight tests being a problem is applicable to our current `{datetime.now}` implementation as well. May I suggest fixing this in another Issue?

HirdayGupta commented:
@whipermr5 Build is now green! (from https://github.com/TEAMMATES/teammates/pull/7715/commits/caf2811b17102e66f51025758592b2a9de313913, only changed documentation in 88a6778). Ready for review! :)

whipermr5 commented:
> There is definitely a difference between running the tests using ./gradlew appengineRun ciTests and running them from Eclipse. Just experienced once more where the test files were updated with the wrong time values when I used the Terminal method.

Can you open an issue and we can continue troubleshooting there? I might have come across the problem myself too but it went away after a clean build I think, and I haven't been able to reproduce it since.

> Yep. Will try this and let you know soon. However, your point about the close to midnight tests being a problem is applicable to our current {datetime.now} implementation as well. May I suggest fixing this in another Issue?

Sure. Did the complete exact match approach work?

whipermr5 commented on src/main/java/teammates/common/util/Const.java:
```diff
@@ -110,6 +110,8 @@
         TIME_REPRESENTS_DEFAULT_TIMESTAMP = TimeHelper.convertToDate("2011-01-01 00:00 AM UTC");
     }
 
+    public static final String TIME_FORMAT_ISO_8601_UTC = "yyyy-MM-dd'T'HH:mm:ss.S'Z'";
```
Apologies I didn't notice this earlier as you'd have to run GodMode again 😅 - could you drop the milliseconds from the timestamp? We don't need that precision and it seems other sites don't use it e.g. GitHub right here (you can inspect the timestamp element of this comment, or any other timestamp on GitHub, to see that they use ISO8601 UTC to carry the timestamp data for formatting in the frontend too, but they don't include milliseconds).

HirdayGupta replied:
@whipermr5 No worries! :-)  Thanks for including such a detailed explanation for the change and a superb example!

HirdayGupta commented:
> Before you do that, maybe just one last troubleshooting step: what's the version of Jetty displayed if you start the dev server from the command line in the foreground i.e. ./gradlew appengineRun -Pdisable_daemon?

@whipermr5 The version displayed is `jetty-6.1.x`.

> Did the complete exact match approach work?

Nope, it did not. It's super unreliable and depends too much on machine configurations. Travis builds run quite fast so it _might_ just work there, but on my local machine, there was sometimes a difference of a minute. (Of course, sometimes the difference between one minute and the next can be just one second!) See #5578.

HirdayGupta commented:
@whipermr5 Ready for review! :)

(Travis build should pass if you restart it :p)

An additional note: We might want to start looking into why `AdminActivityLogPageActionTest` and `AdminEmailLogPageActionTest` fail so often. Usually these failures were confined to test tries 1 and 2, but recently they've been causing build failures too.

whipermr5 commented:
> The version displayed is jetty-6.1.x.

Hmm; can you paste the full output of `./gradlew appengineRun -Pdisable_daemon`?

> An additional note: We might want to start looking into why AdminActivityLogPageActionTest and AdminEmailLogPageActionTest fail so often. Usually these failures were confined to test tries 1 and 2, but recently they've been causing build failures too.

Yup working on it :)

damithc commented:
@HirdayGupta can post a screenshot?

HirdayGupta commented:
Sure @damithc. As requested: 
<img width="1264" alt="screen shot 2017-07-24 at 2 50 58 pm" src="https://user-images.githubusercontent.com/18088721/28516692-b6542038-707f-11e7-8ab0-a94c8f9f7638.png">

HirdayGupta commented:
@damithc @whipermr5 May I resolve the merge conflict?

whipermr5 commented:
@HirdayGupta You have to hahaha it's not an option

HirdayGupta commented:
> You have to hahaha it's not an option

My bad 😆. I was wondering if I would do it or if you all would handle it. :p

whipermr5 commented:
@HirdayGupta Pending a GodMode run of `InstructorFeedbackSubmitPageUiTest` to update `instructorFeedbackSubmitPagePrivate.html`.

whipermr5 commented on src/main/java/teammates/common/util/TimeHelper.java:
```diff
@@ -266,6 +266,17 @@ public static String formatDateTimeForInstructorHomePage(Date date) {
+            return "";
+        }
+        SimpleDateFormat sdf = new SimpleDateFormat(Const.TIME_FORMAT_ISO_8601_UTC);
+        return sdf.format(dateInUtc);
```
I found the issue! You need to do `sdf.setTimeZone(SystemParams.TIME_ZONE);` for it to work without the `user.timezone` VM flag.

HirdayGupta replied:
Great catch! 🎉 

Just wondering, how come tests pass on Travis without this?

whipermr5 replied:
Travis runs on UTC. This system (#7521 and the setting of time zone for every `SimpleDateFormat` instance) is in place for ease of local testing on systems that don't run UTC, and has no effect on systems that are already on UTC.

HirdayGupta replied:
Ahh, I see. Understood. Thanks! 👍
